flocker/provision/_digitalocean.py

```diff
@@ -0,0 +1,370 @@
+
+    :param error_checkers: A ``list`` of ``callables`` which will check for
+        expected exceptions.
+    :param callable: The API function to call.
+    :param args: Positional arguments to supply when calling it.
+    :param kwargs: Keyword arguments to supply when calling it.
+    :return: The result of calling  ``callable``.
+    """
+    while True:
+        try:
+            result = callable(*args, **kwargs)
+        except Exception as e:
+            for checker in error_checkers:
+                if checker(e):
+                    time.sleep(1)
+                    break
+            else:
+                raise
+        else:
+            return result
+
+
+def pending_event(exception):
+    """
+    Check for a pending event exception.
+
+    DigitalOceanV2 API only allows one change at a time and returns HTTP code
+    402 if another change is already pending.
+
+    So this function repeats the API call if that error code is received and
+    returns the result if the call eventually succeeds.
+
+    The raw DO API returns ``event``s whose status can be queried, and that
+    would be a better way to block before issuing the next API call, but
+    pyocean doesn't consistently return the event info. E.g. droplet.create
+    returns a ``droplet`` instance instead whose status is difficult to check.
+
+    See https://digitalocean.uservoice.com/forums/136585-digitalocean/suggestions/4842992-allow-api-cal # noqa
+    """
+    # Import here, so that this can be added to ``flocker.provision`` without
+    # having to install ``pyocean``.
+    import pyocean
+
+    if isinstance(exception, pyocean.exceptions.ClientError):
+        if exception.message == 'Droplet already has a pending event.':
+            return True
+    return False
+
+
+def droplet_still_on(exception):
+    """
+    Check for a droplet still on exception.
+
+    Shutdown returns the following event, indicating that the
+    droplet has halted.
+
+    {u'completed_at': u'2015-01-15T20:52:36Z',
+     u'id': 41364967,
+     u'region': u'ams3',
+     u'resource_id': 3797602,
+     u'resource_type': u'droplet',
+     u'started_at': u'2015-01-15T20:52:31Z',
+     u'status': u'completed',
+     u'type': u'shutdown'}
+
+    But it still seems to require some time before powering on, so catch the
+    "currently on" exception and retry in that case.
+    """
+    # Import here, so that this can be added to ``flocker.provision`` without
+    # having to install ``pyocean``.
+    import pyocean
+
+    if (isinstance(exception, pyocean.exceptions.ClientError)
+        and exception.message == ('Droplet is currently on. '
+                                  'Please power it off to run this event.')):
+        return True
+    return False
+
+
+def retry_if_pending(callable, *args, **kwargs):
+    """
+    Repeats the API call if another API event is currently in progress.
+
+    It returns the result if the call eventually succeeds.
```

Comment:
Yeah, `trial` will eventually halt the test, but it doesn't seem appropriate to mention that here....as discussed.